src/System.Private.CoreLib/shared/System/TimeZoneInfo.Unix.cs

```diff
@@ -416,21 +417,29 @@ private static IEnumerable<string> EnumerateFilesRecursively(string path)
             string currentPath = path;
             for(;;)
             {
-                using (Microsoft.Win32.SafeHandles.SafeDirectoryHandle dirHandle = Interop.Sys.OpenDir(currentPath))
+                IntPtr dirHandle = Interop.Sys.OpenDir(currentPath);
+                if (dirHandle == IntPtr.Zero)
                 {
-                    if (dirHandle.IsInvalid)
-                    {
-                        throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), currentPath, isDirectory: true);
-                    }
+                    throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), currentPath, isDirectory: true);
+                }
+
+                byte[] dirBuffer = null;
+                try
+                {
+                    dirBuffer = ArrayPool<byte>.Shared.Rent(Interop.Sys.ReadBufferSize);
 
                     // Read each entry from the enumerator
-                    Interop.Sys.DirectoryEntry dirent;
-                    while (Interop.Sys.ReadDir(dirHandle, out dirent) == 0)
+                    Interop.Sys.DirectoryEntry dirent = default;
+                    while (Interop.Sys.ReadDir(dirHandle, dirBuffer, ref dirent) == 0)
                     {
-                        if (dirent.InodeName == "." || dirent.InodeName == "..")
+                        Span<char> nameBuffer = stackalloc char[256];
```

Comment:
Should this magic 255 / 256 constant be in `Interop.ReadDir.cs` next to `GetName` ?